mono/metadata/threads.c

```diff
@@ -464,35 +502,128 @@ static void ensure_synch_cs_set (MonoInternalThread *thread)
 	synch_cs = g_new0 (MonoCoopMutex, 1);
 	mono_coop_mutex_init_recursive (synch_cs);
 
-	if (mono_atomic_cas_ptr ((gpointer *)&thread->synch_cs,
-					       synch_cs, NULL) != NULL) {
+	gpointer ret = mono_atomic_cas_ptr ((gpointer *)&thread->synch_cs, synch_cs, NULL);
+
+	if (ret != NULL) {
 		/* Another thread must have installed this CS */
+		/* Or the finalizer had to have run */
 		mono_coop_mutex_destroy (synch_cs);
 		g_free (synch_cs);
 	}
 }
 
-static inline void
-lock_thread (MonoInternalThread *thread)
+static void
+internal_thread_free (MonoInternalThread *thread)
+{
+	MonoCoopMutex *synch_cs = thread->synch_cs;
+	thread->synch_cs = INTERNAL_THREAD_TOMBSTONE;
+	mono_memory_barrier ();
+
+	mono_coop_mutex_destroy (synch_cs);
+	g_free (synch_cs);
+}
+
+static void
+cleanup_thread_lock (MonoInternalThread *thread)
+{
+	mono_memory_barrier ();
+	g_assert (sizeof (MonoInternalThreadLockFlags) == sizeof (gint32));
+
+	gint32 flags = mono_atomic_load_i32 (&thread->lock_flags.placeholder);
+	MonoInternalThreadLockFlags *rich_flags = (MonoInternalThreadLockFlags *) &flags;
+
```

Comment:
No.
```
MonoInternalLockFlags flags;
flags.i = mono_atomic_load_i32(&thread->lock_flags.i);
```

Reply:
That's a really awkward interface. I guess I can do it like that.

Reply:
It is not more awkward than pointers and casting.

Reply:
Casting is visually a side-effect-free operation while assigning to locals to me feels like it should be more than just a value unwrapping step. If this were going to be used elsewhere, I would wrap it with a macro or something, but this change only impacts these 3 functions so it's fine.

Reply:
Picture how too many of these in a large method would completely destroy the ability of someone quickly scanning through the code to figure out what inputs and external state determines the control flow

Reply:
There is union and atomic operations, locks, and threads, plenty of warning signs.